File: clients/macos/Sources/FuseOSCore/ClipHistoryStore.swift
import Foundation

/// Keeps clipboard history across app launches.
///
/// On disk, not in the control plane: clipboard content never reaches the server or its
/// database (see CLAUDE.md). This is local storage on the machine that already has the
/// content on its pasteboard, which is a different thing from shipping it anywhere.
///
/// Text lives in one JSON index; image bytes go to files beside it, referenced by name.
/// Base64-ing a 2 MB screenshot into JSON would triple it and force the whole history to
/// be parsed to read one entry.
public struct ClipHistoryStore {
    private let directory: URL
    private var index: URL { directory.appendingPathComponent("history.json") }
    private var blobs: URL { directory.appendingPathComponent("blobs", isDirectory: true) }

    /// Defaults to Application Support, which is where user data that is not a document
    /// belongs and what Time Machine backs up.
    public init(directory: URL? = nil) {
        self.directory = directory ?? FileManager.default
            .urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
            .appendingPathComponent("FuseOS/clip-history", isDirectory: true)
    }

    private struct Record: Codable {
        var id: Int
        var text: String?
        var mime: String?
        var blob: String?
        var fromSelf: Bool
        var atUnixMs: Int64
    }

    public func load() -> [ClipEntry] {
        guard let data = try? Data(contentsOf: index),
              let records = try? JSONDecoder().decode([Record].self, from: data)
        else { return [] }

        return records.compactMap { record in
            var imageData: Data?
            if let blob = record.blob {
                // An entry whose blob went missing is dropped rather than shown as an
                // empty card the user cannot do anything with.
                guard let bytes = try? Data(contentsOf: blobs.appendingPathComponent(blob)) else {
                    return nil
                }
                imageData = bytes
            }
            return ClipEntry(
                id: record.id,
                text: record.text,
                imageData: imageData,
                mime: record.mime,
                fromSelf: record.fromSelf,
                at: Date(timeIntervalSince1970: Double(record.atUnixMs) / 1000),
            )
        }
    }

    /// Writes the whole list. Small and bounded, so a full rewrite beats incremental edits.
    public func save(_ entries: [ClipEntry]) {
        let fileManager = FileManager.default
        try? fileManager.createDirectory(at: blobs, withIntermediateDirectories: true)

        var keep = Set<String>()
        let records: [Record] = entries.map { entry in
            var blob: String?
            if let data = entry.imageData {
                let name = "\(entry.id).bin"
                try? data.write(to: blobs.appendingPathComponent(name))
                keep.insert(name)
                blob = name
            }
            return Record(
                id: entry.id,
                text: entry.text,
                mime: entry.mime,
                blob: blob,
                fromSelf: entry.fromSelf,
                atUnixMs: Int64(entry.at.timeIntervalSince1970 * 1000),
            )
        }

        if let data = try? JSONEncoder().encode(records) {
            try? data.write(to: index)
        }
        // Evicted entries leave their blobs behind otherwise, and images are the only
        // thing here big enough to matter.
        let existing = (try? fileManager.contentsOfDirectory(atPath: blobs.path)) ?? []
        for name in existing where !keep.contains(name) {
            try? fileManager.removeItem(at: blobs.appendingPathComponent(name))
        }
    }

    /// Sign-out must not leave the last user's copied content on disk.
    public func clear() {
        try? FileManager.default.removeItem(at: directory)
    }
}

/// How far back the history list shows. Mirrors `HistoryWindow` on Android.
public enum HistoryWindow: String, CaseIterable, Identifiable, Sendable {
    case day, week, month, all

    public var id: String { rawValue }

    public var label: String {
        switch self {
        case .day: return "24 hours"
        case .week: return "7 days"
        case .month: return "30 days"
        case .all: return "All"
        }
    }

    /// Nil means everything still held — the caps in `ClipboardSync` bound that, not time.
    var seconds: TimeInterval? {
        switch self {
        case .day: return 24 * 60 * 60
        case .week: return 7 * 24 * 60 * 60
        case .month: return 30 * 24 * 60 * 60
        case .all: return nil
        }
    }

    public func filter(_ entries: [ClipEntry], now: Date = Date()) -> [ClipEntry] {
        guard let seconds else { return entries }
        let cutoff = now.addingTimeInterval(-seconds)
        return entries.filter { $0.at >= cutoff }
    }
}

File: clients/android/app/src/main/java/com/fuseos/app/clipboard/ClipHistoryStore.kt
package com.fuseos.app.clipboard

import android.util.Log
import org.json.JSONArray
import org.json.JSONObject
import java.io.File

/**
 * Keeps clipboard history across app restarts.
 *
 * On disk, not in the control plane: clipboard content never reaches the server or its
 * database (see CLAUDE.md). This is local storage on the device that already has the
 * content on its clipboard, which is a different thing from shipping it anywhere.
 *
 * Text lives in one JSON index; image bytes go to files beside it, referenced by name.
 * Base64-ing a 2 MB screenshot into JSON would triple it and force the whole history to
 * be parsed to read one entry.
 */
class ClipHistoryStore(private val directory: File) {

    private val index = File(directory, "history.json")
    private val blobs = File(directory, "blobs")

    /** Newest first, oldest entries dropped if they no longer fit the caps. */
    fun load(): List<ClipEntry> = runCatching {
        if (!index.exists()) return emptyList()
        val array = JSONArray(index.readText())
        (0 until array.length()).mapNotNull { i ->
            val o = array.getJSONObject(i)
            val blob = o.optString("blob").takeIf { it.isNotEmpty() }
            val bytes = blob?.let { name ->
                File(blobs, name).takeIf { it.exists() }?.readBytes()
                // An entry whose blob went missing is dropped rather than shown as an
                // empty card the user cannot do anything with.
                    ?: return@mapNotNull null
            }
            ClipEntry(
                id = o.getLong("id"),
                text = o.optString("text").takeIf { it.isNotEmpty() },
                imageBytes = bytes,
                mime = o.optString("mime").takeIf { it.isNotEmpty() },
                fromSelf = o.getBoolean("fromSelf"),
                atUnixMs = o.getLong("at"),
            )
        }
    }.onFailure { Log.w(TAG, "history unreadable, starting empty: ${it.message}") }
        .getOrDefault(emptyList())

    /** Writes the whole list. Small and bounded, so a full rewrite beats incremental edits. */
    fun save(entries: List<ClipEntry>) {
        runCatching {
            directory.mkdirs()
            blobs.mkdirs()
            val array = JSONArray()
            val keep = mutableSetOf<String>()
            for (entry in entries) {
                val o = JSONObject()
                    .put("id", entry.id)
                    .put("fromSelf", entry.fromSelf)
                    .put("at", entry.atUnixMs)
                entry.text?.let { o.put("text", it) }
                entry.mime?.let { o.put("mime", it) }
                entry.imageBytes?.let { bytes ->
                    val name = "${entry.id}.bin"
                    File(blobs, name).writeBytes(bytes)
                    keep += name
                    o.put("blob", name)
                }
                array.put(o)
            }
            index.writeText(array.toString())
            // Evicted entries leave their blobs behind otherwise, and images are the only
            // thing here big enough to matter.
            blobs.listFiles()?.forEach { if (it.name !in keep) it.delete() }
        }.onFailure { Log.w(TAG, "could not save history: ${it.message}") }
    }

    /** Sign-out must not leave the last user's copied content on disk. */
    fun clear() {
        runCatching {
            index.delete()
            blobs.deleteRecursively()
        }
    }

    private companion object {
        const val TAG = "FuseHistory"
    }
}

/** How far back the history list shows. Mirrors `HistoryWindow` on macOS. */
enum class HistoryWindow(val label: String, val millis: Long?) {
    Day("24 hours", 24L * 60 * 60 * 1000),
    Week("7 days", 7L * 24 * 60 * 60 * 1000),
    Month("30 days", 30L * 24 * 60 * 60 * 1000),

    /** Everything still held. The caps in `ClipboardSync` bound this, not time. */
    All("All", null),
    ;

    fun filter(entries: List<ClipEntry>, now: Long = System.currentTimeMillis()): List<ClipEntry> {
        val cutoff = millis?.let { now - it } ?: return entries
        return entries.filter { it.atUnixMs >= cutoff }
    }
}
